### PR TITLE
Add `AbstractRuleBuilder.onGrpcTrailers()`

### PR DESCRIPTION
**Pull request by trustin**
Related issue: #4496

Motivation:

A user might want to write a circuit breaker or retry rule that makes a decision based on gRPC trailers. Accessing gRPC trailers can be tricky because they can be present in either HTTP headers, trailers or even content (if gRPC Web). It would save a lot of duplicate effort to write a filter that checks difference places if we provide a built-in filter for gRPC trailers.

Modifications:

- Add `AbstractRuleBuilder.onGrpcTrailers()`
- Modify `AbstractRuleBuilderUtil` so it handles a gRPC trailers filter.
- Move the core logic of `GrpcWebTrailers` to `InternalGrpcWebTrailers` in the `core` module, because otherwise `AbstractRuleBuilderUtil` cannot access gRPC Web trailers.
- Add `AbstractRuleBuilder.requiresResponseTrailers()` to reduce duplication
- Add `AbstractRuleBuilder.combinePredicates()` to reduce duplication.
- Fix some Javadoc errors.

Result:

- A user can write a circuit breaker or retry rule that makes a decision based on gRPC trailers, regardless of whether it's a gRPC response with/without body or even gRPC Web response.
- Partially implemented #4496 although a little bit different from the initially suggested idea

### CODE REVIEW

**Comment by ikhoon** [comment]
Would you also mind migrating `GrpcClientRetryTest` to use `onGrpcTrailers()`?
https://github.com/line/armeria/blob/8ed438da36ba4e0f7d111c00f748b3fcc2906e2c/grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientRetryTest.java#L80

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java [hunk 1]
```diff
@@ -54,11 +88,17 @@ public final class AbstractRuleBuilderUtil {
...
+                responseTrailersFilter == null && grpcTrailersFilter == null &&
+                !hasResponseFilter) {
                 return true;
             }
 
-            if (cause != null && exceptionFilter != null && exceptionFilter.test(ctx, Exceptions.peel(cause))) {
+            return applySlow(ctx, cause, log);
+        }
+
+        private boolean applySlow(ClientRequestContext ctx, Throwable cause, RequestLog log) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:99]: nit:
```suggestion
        private boolean applySlow(ClientRequestContext ctx, @Nullable Throwable cause, RequestLog log) {
```

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java [hunk 2]
```diff
@@ -41,9 +42,42 @@ public final class AbstractRuleBuilderUtil {
...
+               boolean hasResponseFilter) {
+            this.requestHeadersFilter = requestHeadersFilter;
+            this.exceptionFilter = exceptionFilter;
+            this.responseHeadersFilter = responseHeadersFilter;
+            this.responseTrailersFilter = responseTrailersFilter;
+            this.grpcTrailersFilter = grpcTrailersFilter;
+            this.hasResponseFilter = hasResponseFilter;
+        }
+
+        // TODO(trustin): Figure out why apply() is called more than once for each client call
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:77]: Which test case makes calls this more than once?

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 3]
```diff
@@ -185,6 +174,23 @@ public AbstractRuleBuilder onException() {
         return onException((unused1, unused2) -> true);
     }
 
+    private static <T> BiPredicate<ClientRequestContext, T> combinePredicates(
+            @Nullable
+            BiPredicate<ClientRequestContext, T> firstPredicate,
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:179]: Better to merge this into one line?
Nice deduplication by the way. 👍

File: core/src/test/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientRuleTest.java [hunk 4]
```diff
@@ -107,14 +122,87 @@ void openCircuitWithTrailer() {
...
+    @Test
+    void openCircuitWithGrpc() {
+        final CircuitBreakerRule rule =
+                CircuitBreakerRule.builder()
+                                  .onGrpcTrailers((ctx, trailers) -> trailers.containsInt("grpc-status", 3))
+                                  .thenFailure();
+
+        final BlockingWebClient client =
+                WebClient.builder(server.httpUri())
+                         .decorator(CircuitBreakerClient.newDecorator(CircuitBreaker.ofDefaultName(), rule))
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientRuleTest.java:142]: How about configuring the `CircuitBreaker` to fail fast?

Comment (minwoox) [reply core/src/test/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientRuleTest.java:142]: Could you also add a test that uses `onGrpcTrailers` in RetryingClient, please?

**Review by minwoox — COMMENTED** [review]
Great job!
Left some minor comments. 😄

**Review by jrhee17 — APPROVED** [review]
Looks good to me once comments by @ikhoon and @minwoox are addressed 👍   Thanks @trustin ! 🙇 👍 🚀

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java [hunk 5]
```diff
@@ -76,9 +116,28 @@ public final class AbstractRuleBuilderUtil {
...
+                    if (!maybeGrpcTrailers.contains("grpc-status")) {
+                        // Check gRPC Web trailers lastly, because gRPC Web is the least used protocol
+                        // in reality.
+                        maybeGrpcTrailers = InternalGrpcWebTrailers.get(ctx);
+                    }
+                }
+
+                // Suppressing the inspection rule because we don't want to return false too early.
+                //noinspection RedundantIfStatement
+                if (maybeGrpcTrailers != null && grpcTrailersFilter.test(ctx, maybeGrpcTrailers)) {
```
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java:134]: I guess @ikhoon's will take care of this situation: https://github.com/line/armeria/pull/4808

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 6]
```diff
@@ -86,15 +80,18 @@ public AbstractRuleBuilder onResponseHeaders(
...
+        return this;
+    }
+
+    /**
+     * Adds the specified {@code grpcTrailersFilter}.
+     */
+    public AbstractRuleBuilder onGrpcTrailers(
+            BiPredicate<? super ClientRequestContext, ? super HttpHeaders> grpcTrailersFilter) {
+        this.grpcTrailersFilter = combinePredicates(this.grpcTrailersFilter, grpcTrailersFilter,
+                                                    "grpcTrailersFilter");
```
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:94]: Fixed. Thanks for the heads up!

File: core/src/test/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientRuleTest.java (hunk 4 above)
Comment (trustin) [reply core/src/test/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientRuleTest.java]: That's a good idea. I added a new factory method `superSensitiveCircuitBreaker()` and used it for all tests in `CircuitBreakerClientRuleTest`.

I also modified `GrpcClientRetryTest` to use `onGrpcTrailers()`, which resolves your second comment.

File: core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java (hunk 2 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java]: I investigated again and it was actually that the client was making two calls, i.e. there's no problem actually. Removing the comment.

**Review by ikhoon — APPROVED** [review]
Thanks, @trustin! 👍

**Review by minwoox — APPROVED** [review]
Thanks! 👍 👍 👍

**Pull request merged**